src/main/java/robot/subsystems/drivetrain/Drivetrain.java

```diff
@@ -34,6 +46,93 @@ public Drivetrain() {
         rightMaster.configPeakCurrentLimit(MAX_CURRENT);
     }
 
+    public void periodic() {
+        autoShift();
+        lastLeftVelocity = getLeftVelocity();
+        lastRightVelocity = getRightVelocity();
+    }
+
+    /**
+     * checks if the shifter can shift and chooses to which gear to shift
+     */
+    public void autoShift() {
+        if (canShift()) {
+            if ((kickDown() || coastDown()) && !solenoid.get())
+                shiftDown();
+            else if (canShiftUp() && solenoid.get())
+                shiftUp();
+        }
+    }
+
+    /**
+     * if the shifter can shift
+     * @return if the robot is not turning and the cool down time is greater than a threshold value
+     */
+    private boolean canShift() {
+        return !isTurning() && (coolDown.get() >= COOLDOWN_TIME || !hasShifted);
+    }
+
+    /**
+     * @return if the robot is turning
+     */
+    private boolean isTurning() {
+        return getLeftVelocity() - getRightVelocity() < TURN_THRESHOLD;
+    }
+
+    /**
+     * if the robot is in kickdown state
+     * @return if the velocity is less than a threshold value and the robot is decelerating
+     * and the current of the motors is correct
+     */
+    private boolean kickDown() {
+        return getRightVelocity() > 0 && getRightVelocity() < KICKDOWN_VELOCITY_THRESHOLD
+                && getRightAcceleration() < KICKDOWN_ACCEL_THRESHOLD
```

Comment:
I Suggest that you use the right velocity as an absolute value and the same for the acceleration

Reply:
And what about the left side?